mlflow/langchain/langchain_tracer.py

```diff
@@ -376,19 +377,28 @@ def on_tool_start(
         """Start span for a tool run."""
         if metadata:
             kwargs.update({"metadata": metadata})
+
+        # For function calling, input_str can be a stringified dictionary
+        # like "{'key': 'value'}". We try parsing it for better rendering,
+        # but conservatively fallback to original if it fails.
+        try:
+            inputs = ast.literal_eval(input_str)
+        except Exception:
+            inputs = input_str
+
         self._start_span(
             span_name=name or self._assign_span_name(serialized, "tool"),
             parent_run_id=parent_run_id,
             span_type=SpanType.TOOL,
             run_id=run_id,
-            inputs=input_str,
+            inputs=inputs,
             attributes=kwargs,
         )
 
     def on_tool_end(self, output: Any, *, run_id: UUID, **kwargs: Any):
         """Run when tool ends running."""
         tool_span = self._get_span_by_run_id(run_id)
-        self._end_span(run_id, tool_span, outputs=str(output))
+        self._end_span(run_id, tool_span, outputs=output)
```

Comment:
Is it safe to remove string conversion here? output seems can be anything

Reply:
Yes, outputs are serialized by `TraceJSONEncoder` when storing as attribute, which runs `str()` anyway.